### PR TITLE
MUI-Datatables: Update to let text labels and download options to be partials.

### PR DESCRIPTION
**Pull request by jdetle**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/gregnb/mui-datatables/blob/master/src/MUIDataTable.js#L285

Its deep merging defaults with provided options so it is natural to make this a partial.

### CODE REVIEW

File: types/mui-datatables/index.d.ts [hunk 1]
```diff
@@ -179,15 +186,15 @@ export interface MUIDataTableOptions {
...
     ) => React.ReactNode;
     disableToolbarSelect?: boolean;
     download?: boolean;
-    downloadOptions?: {
+    downloadOptions?: Partial<{
         filename: string;
         separator: string;
-        filterOptions?: { useDisplayedColumnsOnly: boolean; useDisplayedRowsOnly: boolean };
-    };
+        filterOptions: { useDisplayedColumnsOnly?: boolean; useDisplayedRowsOnly?: boolean };
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
        filterOptions: Partial<{ useDisplayedColumnsOnly: boolean; useDisplayedRowsOnly: boolean }>;
```

If we're going to use the Partial helper type anyway

File: types/mui-datatables/index.d.ts [hunk 2]
```diff
@@ -179,15 +186,15 @@ export interface MUIDataTableOptions {
             lookup: { [key: number]: boolean };
         },
         displayData: Array<{ data: any[]; dataIndex: number }>,
-        setSelectedRows: (rows: number[]) => void
+        setSelectedRows: (rows: number[]) => void,
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
        setSelectedRows: (rows: number[]) => void
```

File: types/mui-datatables/index.d.ts [hunk 3]
```diff
@@ -161,15 +168,15 @@ export interface MUIDataTableOptions {
...
+        changePage: number,
     ) => React.ReactNode;
     customRowRender?: (data: any[], dataIndex: number, rowIndex: number) => React.ReactNode;
     customSearch?: (searchQuery: string, currentRow: any[], columns: any[]) => boolean;
     customSearchRender?: (
         searchText: string,
         handleSearch: any,
         hideSearch: any,
-        options: any
+        options: any,
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
        options: any
```

File: types/mui-datatables/index.d.ts [hunk 4]
```diff
@@ -148,7 +155,7 @@ export interface MUIDataTableIsRowCheck {
         {
             index: number;
             dataIndex: number;
-        }
+        },
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
        }
```

File: types/mui-datatables/index.d.ts [hunk 5]
```diff
@@ -120,8 +120,15 @@ export interface MUIDataTableColumnState extends MUIDataTableColumnOptions {
...
-    customHeadRender?: (columnMeta: MUIDataTableCustomHeadRenderer, updateDirection: (params: any) => any) => string | React.ReactNode;
+    customBodyRender?: (
+        value: any,
+        tableMeta: MUIDataTableMeta,
+        updateValue: (s: any, c: any, p: any) => any,
+    ) => string | React.ReactNode;
+    customHeadRender?: (
+        columnMeta: MUIDataTableCustomHeadRenderer,
+        updateDirection: (params: any) => any,
+    ) => string | React.ReactNode;
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
    customBodyRender?: (value: any, tableMeta: MUIDataTableMeta, updateValue: (s: any, c: any, p: any) => any) => string | React.ReactNode;
    customHeadRender?: (columnMeta: MUIDataTableCustomHeadRenderer, updateDirection: (params: any) => any) => string | React.ReactNode;
```

File: types/mui-datatables/index.d.ts [hunk 6]
```diff
@@ -31,7 +31,7 @@ export interface MUIDataTableState {
     columns: MUIDataTableColumnState[];
     count: number;
     data: any[];
-    displayData: Array<{dataIndex: number; data: any[]}>;
+    displayData: Array<{ dataIndex: number; data: any[] }>;
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
    displayData: Array<{dataIndex: number; data: any[]}>;
```

File: types/mui-datatables/index.d.ts [hunk 7]
```diff
@@ -202,7 +209,7 @@ export interface MUIDataTableOptions {
     isRowSelectable?: (dataIndex: number, selectedRows?: MUIDataTableIsRowCheck) => boolean;
     onCellClick?: (
         colData: any,
-        cellMeta: { colIndex: number; rowIndex: number; dataIndex: number; event: React.MouseEvent }
+        cellMeta: { colIndex: number; rowIndex: number; dataIndex: number; event: React.MouseEvent },
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
        cellMeta: { colIndex: number; rowIndex: number; dataIndex: number; event: React.MouseEvent }
```

**Review by favna — CHANGES_REQUESTED** [review]
Just a tiny change extra and revert the auto-formatter changes (this will probably also fix the Travis build as it removes the trailing comma).

**Comment by karlasantos** [comment]
Hi! I have been trying to use customFooter of MUIDataTableOptions and looks different of usage in package mui-datatables since v2.9.0. Is there any chance of mapping the changes in @types/mui-datatables? 
Sorry for my english.

**Comment by Favna** [comment]
@karlasantos Everyone can contribute to this repository as they see fit and I recommend you do just that if you want changes to be made to the typings information.

**Comment by karlasantos** [comment]
@Favna okay, I will do that, thanks.

**Comment by jdetle** [comment]
I’ll get on it this weekend, thanks

**Comment by jdetle** [comment]
@favna thanks for the review suggestions, saved me a bit of time. That work is done.

File: types/mui-datatables/index.d.ts [hunk 8]
```diff
@@ -183,11 +183,11 @@ export interface MUIDataTableOptions {
     ) => React.ReactNode;
     disableToolbarSelect?: boolean;
     download?: boolean;
-    downloadOptions?: {
+    downloadOptions?: Partial<{
         filename: string;
         separator: string;
-        filterOptions?: { useDisplayedColumnsOnly: boolean; useDisplayedRowsOnly: boolean };
-    };
+        filterOptions: Partial<{ useDisplayedColumnsOnly: boolean; useDisplayedRowsOnly?: boolean }>;
```
Comment (Favna) [inline types/mui-datatables/index.d.ts]:
```suggestion
        filterOptions: Partial<{ useDisplayedColumnsOnly: boolean; useDisplayedRowsOnly: boolean }>;
```

Comment (Favna) [reply types/mui-datatables/index.d.ts]: The `?` is not needed since it's already `Partial<>`

**Review by favna — CHANGES_REQUESTED** [review]
Just 1 small one left

**Review by favna — APPROVED** [review]

**Comment by jdetle** [comment]
@typescript-bot can you remove the revision needed label?

**Comment by Favna** [comment]
@orta been 5 days since the approval, tagging you since it would be nice if it could get merged.

**Comment by orta** [comment]
Cool, seems good to me

**Pull request merged**